ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,2157 @@
+      NativeType::Bool => fast_api::Type::Bool,
+      NativeType::U8 | NativeType::U16 | NativeType::U32 => {
+        fast_api::Type::Uint32
+      }
+      NativeType::I8 | NativeType::I16 | NativeType::I32 => {
+        fast_api::Type::Int32
+      }
+      NativeType::F32 => fast_api::Type::Float32,
+      NativeType::F64 => fast_api::Type::Float64,
+      NativeType::Void => fast_api::Type::Void,
+      NativeType::I64 => fast_api::Type::Int64,
+      NativeType::U64 => fast_api::Type::Uint64,
+      NativeType::ISize => fast_api::Type::Int64,
+      NativeType::USize | NativeType::Pointer | NativeType::Function => {
+        fast_api::Type::Uint64
+      }
+      NativeType::Buffer => fast_api::Type::TypedArray(fast_api::CType::Uint8),
+    }
+  }
+}
+
+macro_rules! x64 {
+  ($assembler:expr; $($tokens:tt)+) => {
+    dynasm!($assembler; .arch x64; $($tokens)+)
+  }
+}
+
+macro_rules! aarch64 {
+  ($assembler:expr; $($tokens:tt)+) => {
+    dynasm!($assembler; .arch aarch64; $($tokens)+)
+  }
+}
+
+struct SysVAmd64 {
+  // Reference: https://refspecs.linuxfoundation.org/elf/x86_64-abi-0.99.pdf
+  assmblr: dynasmrt::x64::Assembler,
+  // Parameter counters
+  integer_params: u32,
+  float_params: u32,
+  // Stack offset accumulators
+  offset_trampoline: u32,
+  offset_callee: u32,
+  allocated_stack: u32,
+  frame_pointer: u32,
+}
+
+#[cfg_attr(
+  not(all(target_aarch = "x86_64", target_family = "unix")),
+  allow(dead_code)
+)]
+impl SysVAmd64 {
+  // Integer arguments go to the following GPR, in order: rdi, rsi, rdx, rcx, r8, r9
+  const INTEGER_REGISTERS: u32 = 6;
+  // SSE arguments go to the first 8 SSE registers: xmm0-xmm7
+  const FLOAT_REGISTERS: u32 = 8;
+
+  fn new() -> Self {
+    Self {
+      assmblr: dynasmrt::x64::Assembler::new().unwrap(),
+      integer_params: 0,
+      float_params: 0,
+      // Start at 8 to account for trampoline caller's return address
+      offset_trampoline: 8,
+      // default to tail-call mode. If a new stack frame is allocated this becomes 0
+      offset_callee: 8,
+      allocated_stack: 0,
+      frame_pointer: 0,
+    }
+  }
+
+  fn compile(sym: &Symbol) -> Trampoline {
+    let mut compiler = Self::new();
+
+    let must_cast_return_value =
+      compiler.must_cast_return_value(sym.result_type);
+    let must_wrap_return_value =
+      compiler.must_wrap_return_value_in_typed_array(sym.result_type);
+    let must_save_preserved_register = must_wrap_return_value;
+    let cannot_tailcall = must_cast_return_value || must_wrap_return_value;
+
+    if cannot_tailcall {
+      if must_save_preserved_register {
+        compiler.save_preserved_register_to_stack();
+      }
+      compiler.allocate_stack(&sym.parameter_types);
+    }
+
+    for param in &sym.parameter_types {
+      compiler.move_left(param)
+    }
+    if !compiler.is_recv_arg_overridden() {
+      // the receiver object should never be expected. Avoid its unexpected or deliberate leak
+      compiler.zero_first_arg();
+    }
```

Comment:
Perhaps not necessary? The receiver has either been shifted out because it was the first integer/pointer argument, or it's inaccessible in the callback because there are only floating point arguments.

Reply:
I believe this was a conscious decision by the implementer:
1. Always have the same kind of code in the `compile()` function, even if some of it is not strictly necessary for that particular architecture. This way all architecture code looks similar.
2. Explicitly zero out the receiver even if there are only floating point arguments, thus making sure that a C API cannot get its grubby hands on the pointer number even if it lies about its function signature.

Reply:
I suppose that makes sense but in the scenario you're describing you'd ideally poison all unused parameter-passing registers, not just the first one.

Reply:
What could the foreign code even do with the JS pointer? Fill it with garbage bytes? Access OOB within the JS memory heap? Okay, that sounds dangerous, but FFI is *really* unsafe anyways, and can simply segfault, or do anything that JS/Deno can do already, unless I'm mistaken?

The foreign code wouldn't have access to the JS context or any of Deno's internals? (Technically, as they are already within the same addressing space, it's possible to access these anyways)

I'm just doubtful that it's much of a concern.